File: integrations/eye-tracking/flutter-runtime/lib/proof/pop_privacy_gate.dart
/// POP proof packet privacy gate — blocks biometric leakage (Stage 9).
library;

const Set<String> kForbiddenProofJsonKeys = {
  'landmarks',
  'fullLandmarks',
  'mesh',
  'yPlane',
  'y8',
  'jpegBytes',
  'frameBytes',
  'rawGazeStream',
  'cameraFrame',
  'bitmap',
};

/// Depth-first search for forbidden keys in nested JSON-like maps.
List<String> findForbiddenProofKeys(
  Object? value, {
  String path = r'$',
}) {
  final hits = <String>[];
  if (value is Map) {
    for (final entry in value.entries) {
      final key = entry.key.toString();
      final childPath = '$path.$key';
      if (kForbiddenProofJsonKeys.contains(key)) {
        hits.add(childPath);
      }
      hits.addAll(findForbiddenProofKeys(entry.value, path: childPath));
    }
  } else if (value is List) {
    for (var i = 0; i < value.length; i++) {
      hits.addAll(findForbiddenProofKeys(value[i], path: '$path[$i]'));
    }
  }
  return hits;
}

bool proofJsonPassesPrivacyGate(Map<String, dynamic> json) =>
    findForbiddenProofKeys(json).isEmpty;

File: app/src/lib/popPrivacyGate.ts
/** Forbidden proof-packet keys — must never leave device (POP Stage 9). */
export const FORBIDDEN_PROOF_KEYS = new Set([
  'landmarks',
  'fullLandmarks',
  'mesh',
  'yPlane',
  'y8',
  'jpegBytes',
  'frameBytes',
  'rawGazeStream',
  'cameraFrame',
  'bitmap',
])

export function findForbiddenProofKeys(
  value: unknown,
  path = '$',
): string[] {
  const hits: string[] = []
  if (value == null || typeof value !== 'object') return hits
  if (Array.isArray(value)) {
    value.forEach((item, i) => {
      hits.push(...findForbiddenProofKeys(item, `${path}[${i}]`))
    })
    return hits
  }
  for (const [key, child] of Object.entries(value as Record<string, unknown>)) {
    const childPath = `${path}.${key}`
    if (FORBIDDEN_PROOF_KEYS.has(key)) hits.push(childPath)
    hits.push(...findForbiddenProofKeys(child, childPath))
  }
  return hits
}

export function proofJsonPassesPrivacyGate(json: Record<string, unknown>): boolean {
  return findForbiddenProofKeys(json).length === 0
}

/** Strip web-vision point hints; keep derived aggregates only. */
export function sanitizeEyeTrackingForProof(
  eyeTracking: Record<string, unknown>,
): Record<string, unknown> {
  const { gazePosition: _gaze, landmarks: _lm, ...rest } = eyeTracking
  return {
    ...rest,
    derivedOnly: true,
  }
}
